R/animint.R

```diff
@@ -22,11 +22,76 @@ parsePlot <- function(meta){
       plot.meta$scales[[sc$aesthetics]] <- makeScale(sc)
     }
   }
+  
+  ## Export axis specification as a combination of breaks and
+  ## labels, on the relevant axis scale (i.e. so that it can
+  ## be passed into d3 on the x axis scale instead of on the
+  ## grid 0-1 scale). This allows transformations to be used
+  ## out of the box, with no additional d3 coding.
+  theme.pars <- ggplot2:::plot_theme(meta$plot)
+  
+  ## No legend if theme(legend.postion="none").
+  plot.meta$legend <- if(theme.pars$legend.position != "none"){
+    getLegendList(meta$built)
+  }
+  
+  ## save each layer
   for(layer.i in seq_along(meta$plot$layers)){
     ##cat(sprintf("%4d / %4d layers\n", layer.i, length(meta$plot$layers)))
 
     ## This is the layer from the original ggplot object.
     L <- meta$plot$layers[[layer.i]]
+    
+    ## If any legends are specified, add showSelected aesthetic
+    for(legend.i in seq_along(plot.meta$legend)) {
+      # the type of legend: colour, fill, etc
+      legend_type <- plot.meta$legend[[legend.i]]$legend_type
+      # the name of the variable used in this legend
+      var_name <- plot.meta$legend[[legend.i]]$vars
+      # the actual values for that variable
+      var <- L$data[[var_name]]
+      ## checking if it is a discrete variable.
+      if(plyr::is.discrete(var)) {
+        is.interactive.aes <-
+          grepl("showSelected|clickSelects", names(L$mapping))
+        is.legend.var <- L$mapping == var_name
+        ## If var_name is used with another interactive aes, then do
+        ## not add any showSelected aesthetic for it.
+        var.is.interactive <- any(is.interactive.aes & is.legend.var)
+        if(!var.is.interactive){
+          for(i in legend_type) {
+            temp_name <- paste0("showSelectedlegend", i)
+            L$mapping[[temp_name]] <- as.symbol(var_name)
+          }
+        }
+        # if selector.types is not specified, set it to multiple
+        if(is.null(meta$selector.types[[var_name]])) {
+          meta$selector.types[[var_name]] <- "multiple"
+        }
```

Comment:
@tdhock Would it make sense to coerce `meta$selector.types` to a list?  I'm getting an error when trying to run this example.

```
mtcars$cyl <- factor(mtcars$cyl)
mtcars$vs <- factor(mtcars$vs)
p1 <- ggplot() + 
  geom_point(aes(wt, mpg, colour = cyl, size = vs), 
             data = mtcars)
p2 <- ggplot() + 
  geom_point(aes(wt, drat, colour = cyl, size = vs, clickSelects = vs), 
             data = mtcars)
animint2dir(list(p1 = p2, p2 = p2))
```